### PR TITLE
update: always rescue formula loading

### PR DESCRIPTION
**Pull request by xu-cheng**
Short term fix

### CODE REVIEW

**Comment by bfontaine** [comment]
You might want to add a comment above these lines explaining it’s a short-term fix.

**Comment by UniqMartin** [comment]
I'd be curious to know what this is supposed to fix and what the long-term solution will be.

**Comment by xu-cheng** [comment]
By short term fix, I mean we should probably reimplement `brew update` rather than this is a fix which will be reverted soon.

**Comment by xu-cheng** [comment]
@bfontaine @UniqMartin see #45616

**Comment by bfontaine** [comment]
@xu-cheng Yeah I’m aware of this issue :wink: but it might be worth adding a link to it as a comment above these `rescue Exception` just in case someone reads this code and wonder why we’re rescuing all Exceptions.

**Comment by UniqMartin** [comment]
I wasn't aware, so thanks for the issue link! I agree that there should be a comment mentioning the issue number and that it's a short-term fix, though.

**Comment by xu-cheng** [comment]
Updated

**Comment by UniqMartin** [comment]
LGTM :+1:

**Comment by bfontaine** [comment]
Thank you @xu-cheng :heart:

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -108,7 +108,8 @@ def update
 
       begin
         f = Formulary.factory("#{user}/#{repo}/#{newname}")
-      rescue FormulaUnavailableError, *FormulaVersions::IGNORED_EXCEPTIONS
+      # short term fix to prevent situation like https://github.com/Homebrew/homebrew/issues/45616
+      rescue Exception
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: Should probably also handle `RuntimeError` with just a `rescue` if you go this route?

**Comment by mikemcquaid** [comment]
The problem with this is if there's an error there's no indication to the user anything has happened and they will just get incorrect output. The longer-term fix, I think, is not ever to load formula code in the `update` process but instead shell out to a new process _after_ update.

File: Library/Homebrew/cmd/update.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: `Exception` will cover them all.

**Comment by xu-cheng** [comment]
I'm exploring the long term fix. No ETA though.

**Comment by UniqMartin** [comment]
Definitely agree on the long-term fix. `brew update` should really be a two-step process that first updates all the Git repositories and then spawns a new process that runs the updated code to generate the report and to do the renames, migrations, etc.

I would also be interested in working on that, but I think this will be in a few weeks at the earliest. Would it make sense to create an issue now to coordinate this and to keep everyone interested up-to-date? Whoever happens to start working on this first could post there to save others from duplicating work.

**As for detecting these kinds of problems with `brew update-test`:**

I haven't looked at the code, but I believe it tests the update `origin/master` :arrow_right: `origin/master+PR`. If so, would it be possible to add additional test cases like

- `origin/master` from a week ago :arrow_right: `origin/master+PR`
- `origin/master` from a month ago :arrow_right: `origin/master+PR`
- `origin/master` from three months ago :arrow_right: `origin/master+PR`

without too much effort?

**Comment by mikemcquaid** [comment]
I'm OK with having a tracking issue and will probably be able to start in the next week or so. I don't think additional testing really helps us here; by the time we detect it it's already too late. I'd actually like to remove the `update-test` as it's relatively slow and I think it's lulled us into a false sense of security.

> I'm exploring the long term fix. No ETA though.

If you disagree with the approaches we've said here, please say so. I'm a little frustrated that this fix was pushed through without much time for review while I and other maintainers were asleep.

**Comment by xu-cheng** [comment]
> I'm exploring the long term fix. No ETA though.

I mean I'm exploring the approach we talk about, i.e. reimplement it using bash.  

> I'm a little frustrated that this fix was pushed through without much time for review while I and other maintainers were asleep.

Sorry about that.

**Comment by mikemcquaid** [comment]
> I mean I'm exploring the approach we talk about, i.e. reimplement it using bash.

Great, thanks :+1: 

> Sorry about that.

That's ok :heart:. I understand why it happened and I've done the same in the past but I just figured it was better to say if I was a little :angry: rather than you not know. :bow:

**Pull request closed without merging**